### PR TITLE
KAFKA-14133: Replace EasyMock with Mockito in streams tests

### PR DESCRIPTION
**Pull request by clolov**
Batch 5 of the tests detailed in https://issues.apache.org/jira/browse/KAFKA-14133 which use EasyMock and need to be moved to Mockito.

### CODE REVIEW

**Comment by clolov** [comment]
There were stubbings which were no longer on the call path so I have removed them. The way I checked that I wasn't changing the test behaviour was to use EasyMock.verify on the mocks and confirming that the stubbings were indeed unused prior to my change. There are multiple possibilities for refactoring, but I chose to keep the changes as close to the EasyMock implementation as the PR is already big.

**Comment by clolov** [comment]
I am aware that there are merge conflicts and I will aim to address them over the coming days.

**Comment by clolov** [comment]
@guozhangwang and @cadonna for visibility

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 1]
```diff
@@ -325,25 +318,21 @@ public void shouldHandleRemovedTasksToRecycleFromStateUpdater() {
...
 
-        Mockito.verify(task00Converted).initializeIfNeeded();
-        Mockito.verify(task01Converted).initializeIfNeeded();
-        Mockito.verify(stateUpdater).add(task00Converted);
-        Mockito.verify(stateUpdater).add(task01Converted);
+        verify(task00Converted).initializeIfNeeded();
+        verify(task01Converted).initializeIfNeeded();
+        verify(stateUpdater).add(task00Converted);
+        verify(stateUpdater).add(task01Converted);
+        verify(activeTaskCreator).closeAndRemoveTaskProducerIfNeeded(any());
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:776]: How to specify that we expect this function call only once? Should we use `verify(activeTaskCreator, times(1)).func();` instead?

Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:776]: Ditto elsewhere for replacing `expectLastCall().once()`?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 2]
```diff
@@ -1300,28 +1229,16 @@ public void shouldCloseActiveTasksWhenHandlingLostTasks() throws Exception {
...
-        expectLastCall().anyTimes();
+        when(activeTaskCreator.createTasks(any(), eq(taskId00Assignment))).thenReturn(singletonList(task00));
+        when(standbyTaskCreator.createTasks(eq(taskId01Assignment))).thenReturn(singletonList(task01));
 
+        // The second attempt will return empty tasks.
         makeTaskFolders(taskId00.toString(), taskId01.toString());
         expectLockObtainedFor(taskId00, taskId01);
 
-        // The second attempt will return empty tasks.
-        makeTaskFolders();
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:1877]: Why we can remove those calls?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 3]
```diff
@@ -4091,21 +3815,20 @@ public void shouldListNotPausedTasks() {
         topologyMetadata.pauseTopology(UNNAMED_TOPOLOGY);
 
         assertEquals(taskManager.notPausedTasks().size(), 0);
+
+        verifyConsumerResumedWithAssignment(consumer);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: Why add this additional verification?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 4]
```diff
@@ -271,11 +265,13 @@ public void shouldClassifyExistingTasksWithStateUpdater() {
...
-        Mockito.verify(stateUpdater).remove(standbyTaskToClose.id());
-        Mockito.verify(stateUpdater).remove(restoringActiveTaskToRecycle.id());
-        Mockito.verify(stateUpdater).remove(restoringActiveTaskToClose.id());
+        verify(stateUpdater).getTasks();
+        verify(stateUpdater).remove(standbyTaskToRecycle.id());
+        verify(stateUpdater).remove(standbyTaskToClose.id());
+        verify(stateUpdater).remove(restoringActiveTaskToRecycle.id());
+        verify(stateUpdater).remove(restoringActiveTaskToClose.id());
+
+        verifyConsumerResumedWithAssignment(consumer);
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: There are several tests where we are adding this verification, could you elaborate a bit why?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 5]
```diff
@@ -2180,29 +2035,16 @@ public void shouldCommitAllActiveTasksThatNeedCommittingOnHandleRevocationWithEo
...
-            .andReturn(singletonList(task10));
+        when(standbyTaskCreator.createTasks(eq(assignmentStandby)))
+            .thenReturn(singletonList(task10));
 
         final ConsumerGroupMetadata groupMetadata = new ConsumerGroupMetadata("appId");
-        expect(consumer.groupMetadata()).andReturn(groupMetadata);
-        producer.commitTransaction(expectedCommittedOffsets, groupMetadata);
-        expectLastCall();
-
-        task00.committedOffsets();
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:2761]: Why we can remove those verifications now?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 6]
```diff
@@ -4068,18 +3796,14 @@ public void shouldConvertStandbyTaskToActiveTask() {
...
         when(activeTask.inputPartitions()).thenReturn(taskId00Partitions);
 
-        expect(activeTaskCreator.createTasks(anyObject(), eq(Collections.emptyMap()))).andReturn(Collections.emptySet());
-        expect(standbyTaskCreator.createTasks(eq(taskId00Assignment))).andReturn(singletonList(standbyTask));
-        expect(activeTaskCreator.createActiveTaskFromStandby(eq(standbyTask), eq(taskId00Partitions), anyObject())).andReturn(activeTask);
-        expect(activeTaskCreator.createTasks(anyObject(), eq(Collections.emptyMap()))).andReturn(Collections.emptySet());
-        expect(standbyTaskCreator.createTasks(eq(Collections.emptyMap()))).andReturn(Collections.emptySet());
-
-        replay(standbyTaskCreator, activeTaskCreator, consumer);
+        when(activeTaskCreator.createTasks(any(), eq(Collections.emptyMap()))).thenReturn(Collections.emptySet());
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:4327]: Here the goals are both verifying those functions should be called, and also return the results when they are called (you can see we did not use `andStubReturn` here). When replacing them with `when` we would not verify anymore right? Ditto elsewhere.

**Review by guozhangwang — COMMENTED** [review]
Thanks @clolov , just a few quick meta thoughts:

1) I think we can remove the `StateMachineTask` class as well since it was used as a mock really to compensate the incapabilities of easymock. While now with Mockito I think we would no longer need it. The only thing I'd need to ask is how to stubbing those state manipulations (e.g. if `setCommitNeeded()` is called, then followed `commitNeeded` should return true).

2) There are a few places where we used `expect.andReturn` which was for both stubbing the return values but as well as expecting those functions to be called, when they are replaced with `when.thenReturn` we would not do verifications afaik, is that right? If yes could we fix them?

cc @cadonna for a second pass.

**Comment by clolov** [comment]
Hey @guozhangwang and thank you for the insightful comments! I will aim to give explanations and amend the PR in the upcoming days.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 1 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:776]: By default it expects the method call to happen only once:
```
public static <T> T verify​(T mock)

Verifies certain behavior happened once.
```
https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#verify(T)

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 2 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:1877]: I have altered the makeTaskFolders function to take the second call into account.

Before:
```
expect(stateDirectory.listNonEmptyTaskDirectories()).andReturn(taskFolders).once();
```
After:
```
when(stateDirectory.listNonEmptyTaskDirectories()).thenReturn(taskFolders).thenReturn(Collections.emptyList());
```

As far as I understand subsequent calls for stubbing in Mockito overwrite previous ones, so I cannot do exactly what was done with EasyMock.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 3 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: The below method set an expectation for a method call which was not stubbed:
```
private static void expectRestoreToBeCompleted(final Consumer<byte[], byte[]> consumer,
                                                   final ChangelogReader changeLogReader,
                                                   final boolean changeLogUpdateRequired) {
        final Set<TopicPartition> assignment = singleton(new TopicPartition("assignment", 0));
        expect(consumer.assignment()).andReturn(assignment);
        consumer.resume(assignment); <--- THIS IS THE EXPECTATION
        expectLastCall();
        expect(changeLogReader.completedChangelogs()).andReturn(emptySet()).times(changeLogUpdateRequired ? 1 : 0, 1);
}
```
I wanted to keep this verification so I moved it to a separate function.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 4 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: Same as above.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 5 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:2761]: As far as I am aware Mockito cannot verify interactions with things which are not mocks. This being said, given you suggested I remove the StateMachineTask this might change in subsequent commits.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 6 above)
Comment (clolov) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:4327]: If we use `@RunWith(MockitoJUnitRunner.StrictStubs.class)` and we run the whole test class Mockito verifies that all stubbings have indeed been used.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 1 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:776]: Ack, thanks!

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 3 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java]: Ah I see, thanks!

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 5 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:2761]: Sounds great, thanks.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java (hunk 6 above)
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:4327]: Ack.

**Comment by clolov** [comment]
Polite bump for a review here @cadonna. I have rebased on the latest trunk and I believe I have answered the majority of the comments already on the pull request. Given the size of the change I would prefer to get rid of StateMachineTask in a seprate pull request rather than as part of this one.

**Comment by clolov** [comment]
@cadonna I have rebased this on the latest trunk. Is it possible to get a review? This pull requests is one of the last remaining ones for completing the Mockito migration for streams.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 7]
```diff
@@ -899,65 +841,59 @@ public void shouldSuspendRevokedTaskRemovedFromStateUpdater() {
...
+                .withInputPartitions(taskId02Partitions).build();
         final StreamTask taskToUpdateInputPartitions = statefulTask(taskId03, taskId03ChangelogPartitions)
-            .inState(State.RESTORING)
-            .withInputPartitions(taskId03Partitions).build();
+                .inState(State.RESTORING)
+                .withInputPartitions(taskId03Partitions).build();
         when(stateUpdater.hasRemovedTasks()).thenReturn(true);
         when(stateUpdater.drainRemovedTasks())
-            .thenReturn(mkSet(taskToRecycle0, taskToRecycle1, taskToClose, taskToUpdateInputPartitions));
+                .thenReturn(mkSet(taskToRecycle0, taskToRecycle1, taskToClose, taskToUpdateInputPartitions));
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:868]: nit: adding the indentation is not necessary.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 8]
```diff
@@ -899,65 +841,59 @@ public void shouldSuspendRevokedTaskRemovedFromStateUpdater() {
...
+                .thenReturn(convertedTask0);
+        when(consumer.assignment()).thenReturn(emptySet());
         final TasksRegistry tasks = mock(TasksRegistry.class);
         when(tasks.removePendingTaskToCloseClean(taskToClose.id())).thenReturn(true);
+        when(tasks.removePendingTaskToCloseClean(taskToUpdateInputPartitions.id())).thenReturn(false);
         when(tasks.removePendingTaskToRecycle(taskToRecycle0.id())).thenReturn(taskId00Partitions);
         when(tasks.removePendingTaskToRecycle(taskToRecycle1.id())).thenReturn(taskId01Partitions);
         when(tasks.removePendingTaskToRecycle(
-            argThat(taskId -> !taskId.equals(taskToRecycle0.id()) && !taskId.equals(taskToRecycle1.id())))
+                argThat(taskId -> !taskId.equals(taskToRecycle0.id()) && !taskId.equals(taskToRecycle1.id())))
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:881]: nit: adding indentation not necessary

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 9]
```diff
@@ -899,65 +841,59 @@ public void shouldSuspendRevokedTaskRemovedFromStateUpdater() {
...
-        expectLastCall().times(2);
-        expect(standbyTaskCreator.createStandbyTaskFromActive(eq(taskToRecycle0), eq(taskId00Partitions)))
-            .andStubReturn(convertedTask0);
-        expect(consumer.assignment()).andReturn(emptySet()).anyTimes();
-        consumer.resume(anyObject());
-        expectLastCall().anyTimes();
+        when(activeTaskCreator.createActiveTaskFromStandby(taskToRecycle1, taskId01Partitions, consumer))
+                .thenReturn(convertedTask1);
+        when(standbyTaskCreator.createStandbyTaskFromActive(taskToRecycle0, taskId00Partitions))
+                .thenReturn(convertedTask0);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:873]: nit: we usually indent 4 spaces.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 10]
```diff
@@ -1060,14 +996,12 @@ public void shouldTransitRestoredTaskToRunning() {
         final TasksRegistry tasks = mock(TasksRegistry.class);
         final TaskManager taskManager = setUpTransitionToRunningOfRestoredTask(task, tasks);
         consumer.resume(task.inputPartitions());
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:998]: This needs to become a verification
```java
verify(consumer).resume(task.inputPartitions());
```
further down in the verification block.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 11]
```diff
@@ -1119,17 +1051,15 @@ public void shouldRecycleRestoredTask() {
             .inState(State.CREATED)
             .withInputPartitions(taskId00Partitions).build();
         final TaskManager taskManager = setUpRecycleRestoredTask(statefulTask);
-        expect(standbyTaskCreator.createStandbyTaskFromActive(statefulTask, statefulTask.inputPartitions()))
-            .andStubReturn(standbyTask);
+        when(standbyTaskCreator.createStandbyTaskFromActive(statefulTask, statefulTask.inputPartitions()))
+            .thenReturn(standbyTask);
         activeTaskCreator.closeAndRemoveTaskProducerIfNeeded(statefulTask.id());
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:1056]: This needs to become verification

```java
        verify(activeTaskCreator).closeAndRemoveTaskProducerIfNeeded(statefulTask.id());
```

in the verification block further down.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 12]
```diff
@@ -1193,15 +1119,13 @@ public void shouldCloseCleanRestoredTask() {
         final TasksRegistry tasks = mock(TasksRegistry.class);
         final TaskManager taskManager = setUpCloseCleanRestoredTask(statefulTask, tasks);
         activeTaskCreator.closeAndRemoveTaskProducerIfNeeded(statefulTask.id());
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:1121]: This needs to become verification

```java
verify(activeTaskCreator).closeAndRemoveTaskProducerIfNeeded(statefulTask.id());
```

in the verification block.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 13]
```diff
@@ -1213,37 +1137,33 @@ public void shouldHandleExceptionThrownDuringCloseInCloseCleanRestoredTask() {
         final TaskManager taskManager = setUpCloseCleanRestoredTask(statefulTask, tasks);
         doThrow(RuntimeException.class).when(statefulTask).closeClean();
         activeTaskCreator.closeAndRemoveTaskProducerIfNeeded(statefulTask.id());
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:1139]: This needs to become verification

```java
verify(activeTaskCreator).closeAndRemoveTaskProducerIfNeeded(statefulTask.id());
```

in the verification block.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 14]
```diff
@@ -1268,16 +1188,14 @@ public void shouldCloseDirtyRestoredTask() {
         when(stateUpdater.restoresActiveTasks()).thenReturn(true);
         activeTaskCreator.closeAndRemoveTaskProducerIfNeeded(statefulTask.id());
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:1189]: This needs to become verification

```java
verify(activeTaskCreator).closeAndRemoveTaskProducerIfNeeded(statefulTask.id());
```

in the verification block.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 15]
```diff
@@ -1292,15 +1210,13 @@ public void shouldUpdateInputPartitionsOfRestoredTask() {
         when(stateUpdater.restoresActiveTasks()).thenReturn(true);
         final TaskManager taskManager = setUpTaskManager(ProcessingMode.AT_LEAST_ONCE, tasks, true);
         consumer.resume(statefulTask.inputPartitions());
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:1212]: This needs to become verification

```java
verify(consumer).resume(statefulTask.inputPartitions());
```

in the verification block.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 16]
```diff
@@ -4370,27 +4080,26 @@ private Map<TaskId, StateMachineTask> handleAssignment(final Map<TaskId, Set<Top
...
         }
     }
 
     private static void expectConsumerAssignmentPaused(final Consumer<byte[], byte[]> consumer) {
         final Set<TopicPartition> assignment = singleton(new TopicPartition("assignment", 0));
-        expect(consumer.assignment()).andReturn(assignment);
-        consumer.pause(assignment);
+        when(consumer.assignment()).thenReturn(assignment);
+        doNothing().when(consumer).pause(assignment);
     }
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:4103]: This method is not needed anymore.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 17]
```diff
@@ -1534,66 +1440,61 @@ public void shouldTryToLockValidTaskDirsAtRebalanceStart() throws Exception {
...
+        final Set<TopicPartition> assignment = singleton(new TopicPartition("assignment", 0));
+        when(consumer.assignment()).thenReturn(assignment);
 
         taskManager.handleRebalanceComplete();
         assertThat(taskManager.lockedTaskDirectories(), is(mkSet(taskId00, taskId01)));
-        verify(stateDirectory);
+        verify(consumer).pause(assignment);
+        verify(stateDirectory).unlock(taskId02);
+
+        verifyConsumerResumedWithAssignment(consumer);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:1497]: This method should also take the assignment as an argument. Maybe it is even better to call
```java
verify(consumer, atLeastOnce()).resume(assignment);
```
instead of `verifyConsumerResumedWithAssignment(consumer)` here.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java [hunk 18]
```diff
@@ -1667,6 +1566,8 @@ public void shouldComputeOffsetSumForStandbyTask() throws Exception {
         restoringTask.setChangelogOffsets(changelogOffsets);
 
         assertThat(taskManager.getTaskOffsetSums(), is(expectedOffsetSums));
+
+        verifyConsumerResumedWithAssignment(consumer);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java:1570]: Why did you add this verification? It was not there before the migration, right?

**Review by cadonna — COMMENTED** [review]
@clolov Thanks for your patience!

Here my feedback for the the first 1500  lines. I will continue to review the PR in 1.5 weeks after my vacation.

**Comment by cadonna** [comment]
@clolov Since this PR is rather long and hard to review, I took a stab to subdivide it into smaller PRs by migrating single mocks. I opened a PR for the migration of the topology builder mock: https://github.com/apache/kafka/pull/13529.
Let me know what you think about it. 
If you think that is a good idea feel free to migrate the other mocks. I think we also need a bit to rethink the testing strategy in this test class. Sorry that this migration gets so complicated.

**Comment by clolov** [comment]
Heya @cadonna, thank you for the review! The approach you have taken in the other PR makes sense to me, so I will aim to open a few pull requests in a similar fashion. What worries you about the testing strategy in the class?

**Comment by cadonna** [comment]
> What worries you about the testing strategy in the class?

My worry is that we verify interactions with mocks (mainly strict mocks) in unit tests in which we test unrelated aspects. That makes the test code more complicated than necessary. One example, is [this](https://github.com/apache/kafka/pull/13529#discussion_r1165561113) from the other PR I linked above. In other words, we should probably try not only to migrate to Mockito but also to refactor a bit the unit tests where it makes sense to decrease the verifications on the mocks and thus the complexity of the tests. However, we should try to keep the refactorings at a minimum.

**Comment by clolov** [comment]
Okay, this makes sense to me. I will aim to start opening PRs in the same manner as yours in the upcoming days and let's see where we go!

**Comment by clolov** [comment]
I will close this PR as we have taken a different approach. Please refer to PRs linking to this one in order to find out how we have implemented the changes.
